contrib/format-hdf5/src/main/java/org/apache/drill/exec/store/hdf5/writers/HDF5MapDataWriter.java

```diff
@@ -0,0 +1,136 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.drill.exec.store.hdf5.writers;
+
+import ch.systemsx.cisd.hdf5.HDF5CompoundMemberInformation;
+import ch.systemsx.cisd.hdf5.IHDF5Reader;
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.exec.physical.resultSet.RowSetLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+
+public class HDF5MapDataWriter extends HDF5DataWriter {
+  private static final Logger logger = LoggerFactory.getLogger(HDF5MapDataWriter.class);
+
+  private final String UNSAFE_SPACE_SEPARATOR = " ";
+
+  private final String SAFE_SPACE_SEPARATOR = "_";
+
+  private ArrayList<HDF5DataWriter> dataWriters;
+
+  private ArrayList<String> fieldNames;
+
+
+  public HDF5MapDataWriter(IHDF5Reader reader, RowSetLoader columnWriter, String datapath) {
+    super(reader, columnWriter, datapath);
+    fieldNames = new ArrayList<>();
+
+    compoundData = reader.compounds().readArray(datapath, Object[].class);
+    dataWriters = new ArrayList<>();
+    fieldNames = getFieldNames();
+    try {
+      getDataWriters();
+    } catch (Exception e) {
+      throw UserException.dataReadError().addContext("Error writing Compound Field: ").addContext(e.getMessage()).build(logger);
+    }
+  }
+
+  public boolean write() {
+    if (!hasNext()) {
+      return false;
+    } else {
+      // Loop through the columns and write the columns
+      columnWriter.start();
+      for (int i = 0; i < dataWriters.size(); i++) {
+        dataWriters.get(i).write();
+      }
+      columnWriter.save();
+      counter++;
+      return true;
+    }
+  }
+
+  public boolean hasNext() {
+    return counter < dataWriters.get(0).getDatasize();
+  }
+
+  private ArrayList<String> getFieldNames() {
+    ArrayList<String> names = new ArrayList<>();
+
+    HDF5CompoundMemberInformation[] infos = reader.compounds().getDataSetInfo(datapath);
+    for (HDF5CompoundMemberInformation info : infos) {
+      names.add(info.getName().replace(UNSAFE_SPACE_SEPARATOR, SAFE_SPACE_SEPARATOR));
+    }
+    return names;
+  }
+
+  /**
+   * This function will populate the ArrayList of DataWriters. Since HDF5 Maps contain homogeneous columns,
+   * it is fine to get the first row, and iterate through the columns to get the data types and build the schema accordingly.
+   */
```

Comment:
Do we handle the case where there is no first row?

Reply:
If there was no first row, HDF5 wouldn't allow a map at all.